api/agent/agent.go

```diff
@@ -517,24 +520,28 @@ func (s *hotSlot) exec(ctx context.Context, call *call) error {
 	oldStderr := s.container.swap(call.stderr, &call.Stats)
 	defer s.container.swap(oldStderr, nil) // once we're done, swap out in this scope to prevent races
 
-	errApp := make(chan error, 1)
+	containerError := make(chan error, 1)
+
 	go func() {
-		// TODO make sure stdin / stdout not blocked if container dies or we leak goroutine
-		// we have to make sure this gets shut down or 2 threads will be reading/writing in/out
-		ci := protocol.NewCallInfo(call.Call, call.req)
-		errApp <- s.proto.Dispatch(ctx, ci, call.w)
+		select {
+		case err := <-s.errC: // error from container
+			containerError <- err
+		case <-ctx.Done(): // timeout or cancel from ctx
+			containerError <- ctx.Err()
+			s.done <- ctx.Err()
+		}
 	}()
 
+	ci := protocol.NewCallInfo(call.Call, call.req)
+	dispatchError := s.proto.Dispatch(ci, call.ioWriter)
```

Comment:
the issue here is that i/o could put this over the timeout, so this isn't really an option (sadly). i have a test that can repro this, repro'ing a different issue. 

as far as fixing the race to write to the header bucket -- it seems really easy to slide by making a thread safe `http.ResponseWriter` that we hand into the agent. this would be easily removed once we have fixed the root causal issue (opposed to shuffling things around twice, if the case)

Reply:
we catch the I/O timeout from ctx in the go routine. If dispatch takes forever, then go routine picks up timeout from ctx, then signals container to stop.

Reply:
that makes sense, made this comment before the other (trying to avoid container shutdown, if possible)